awsbestpractices/polling-to-waiters/PollingToWaiters.java

```diff
@@ -0,0 +1,151 @@
+                        new DescribeInstancesRequest().withInstanceIds(instance.getInstanceId()))
+                        .getReservations().get(0).getInstances().get(0);
+            }
+        }
+    }
+
+    // No loop
+    private void waitForTermination_noLoop(AmazonEC2 ec2, List<Instance> terminatedInstances) {
+        ec2.describeInstances(
+                new DescribeInstancesRequest().withInstanceIds(instance.getInstanceId()))
+                .getReservations().get(0).getInstances().get(0);
+        cxt.getClock().sleep(1000);
+    }
+
+    // Can use waiter.
+    // Based on
+    // https://github.com/schibsted/strongbox/blob/fde1861ee83050e8cf82218f212f00ecfa17a003/sdk/src/main/java/com/schibsted/security/strongbox/sdk/internal/kv4j/generic/backend/dynamodb/GenericDynamoDB.java#L130
+    private void waitForTableToBecomeActive(AmazonDynamoDB client) {
+        int retries = 0;
+        String tableStatus = "Unknown";
+        try {
+            while (retries < MAX_RETRIES) {
+                log.info("Waiting for table to become active...");
+                Thread.sleep(SLEEP_TIME);
+                DescribeTableResult result = client.describeTable(tableName);
+                tableStatus = result.getTable().getTableStatus();
+
+                if (tableStatus.equals(TableStatus.ACTIVE.toString()) ||
+                        tableStatus.equals(TableStatus.UPDATING.toString())) {
+                    return;
+                }
+
+                if (tableStatus.equals(TableStatus.DELETING.toString())) {
+                    throw new UnexpectedStateException(
+                            tableName, tableStatus, TableStatus.ACTIVE.toString(),
+                            "Table state changed to 'DELETING' before creation was confirmed");
+                }
+                retries++;
+            }
+        } catch (InterruptedException e) {
+            throw new UnexpectedStateException(tableName, tableStatus, TableStatus.ACTIVE.toString(),
+                    "Error occurred while waiting for DynamoDB table", e);
+        }
+        throw new UnexpectedStateException(tableName, tableStatus, TableStatus.ACTIVE.toString(),
+                "DynamoDB table did not become active before timeout");
+    }
+
+    // Can use waiter.
+    // Based on
+    // https://github.com/xebialabs/overcast/blob/716e57b66870c9afe51edc3b6045863a34fb0061/src/main/java/com/xebialabs/overcast/host/Ec2CloudHost.java#L176
+    // A little controvertial because it does not only wait but also returns value.
+    public String waitUntilRunningAndGetPublicDnsName(AmazonEC2Client ec2) {
+        // Give Amazon some time to settle before we ask it for information
+        sleep(5);
+
+        for (; ; ) {
+            DescribeInstancesRequest describe = new DescribeInstancesRequest().withInstanceIds(asList(instanceId));
+            Instance instance = ec2.describeInstances(describe).getReservations().get(0).getInstances().get(0);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code appears to be waiting for a resource before it runs. You could use the waiters feature to help improve efficiency. Consider using InstanceExists, InstanceRunning, InstanceStopped or InstanceTerminated. For more information, see [https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/](https://aws.amazon.com/blogs/developer/waiters-in-the-aws-sdk-for-java/)